src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1505 @@
+        {
+            return returnStatementAst.Pipeline.Accept(this);
+        }
+
+        object ICustomAstVisitor.VisitExitStatement(ExitStatementAst exitStatementAst)
+        {
+            return TypeInferenceContext.EmptyPSTypeNameArray;
+        }
+
+        object ICustomAstVisitor.VisitThrowStatement(ThrowStatementAst throwStatementAst)
+        {
+            return TypeInferenceContext.EmptyPSTypeNameArray;
+        }
+
+        object ICustomAstVisitor.VisitDoUntilStatement(DoUntilStatementAst doUntilStatementAst)
+        {
+            return doUntilStatementAst.Body.Accept(this);
+        }
+
+        object ICustomAstVisitor.VisitAssignmentStatement(AssignmentStatementAst assignmentStatementAst)
+        {
+            return assignmentStatementAst.Left.Accept(this);
+        }
+
+        object ICustomAstVisitor.VisitPipeline(PipelineAst pipelineAst)
+        {
+            return pipelineAst.PipelineElements.Last().Accept(this);
+        }
+
+        object ICustomAstVisitor.VisitCommand(CommandAst commandAst)
+        {
+            return InferTypesFrom(commandAst);
+        }
+
+        object ICustomAstVisitor.VisitCommandExpression(CommandExpressionAst commandExpressionAst)
+        {
+            return commandExpressionAst.Expression.Accept(this);
+        }
+
+        object ICustomAstVisitor.VisitCommandParameter(CommandParameterAst commandParameterAst)
+        {
+            return TypeInferenceContext.EmptyPSTypeNameArray;
+        }
+
+        object ICustomAstVisitor.VisitFileRedirection(FileRedirectionAst fileRedirectionAst)
+        {
+            return TypeInferenceContext.EmptyPSTypeNameArray;
+        }
+
+        private IEnumerable<PSTypeName> InferTypesFrom(CommandAst commandAst)
+        {
+            PseudoBindingInfo pseudoBinding = new PseudoParameterBinder()
+                                    .DoPseudoParameterBinding(commandAst, null, null, PseudoParameterBinder.BindingType.ParameterCompletion);
+            if (pseudoBinding?.CommandInfo == null)
+            {
+                yield break;
+            }
+
+            AstParameterArgumentPair pathArgument;
+            string pathParameterName = "Path";
+            if (!pseudoBinding.BoundArguments.TryGetValue(pathParameterName, out pathArgument))
+            {
+                pathParameterName = "LiteralPath";
+                pseudoBinding.BoundArguments.TryGetValue(pathParameterName, out pathArgument);
+            }
+
+            var commandInfo = pseudoBinding.CommandInfo;
+            var pathArgumentPair = pathArgument as AstPair;
+            if (pathArgumentPair?.Argument is StringConstantExpressionAst)
+            {
+                var pathValue = ((StringConstantExpressionAst)pathArgumentPair.Argument).Value;
+                try
+                {
+                    commandInfo = commandInfo.CreateGetCommandCopy(new object[] { "-" + pathParameterName, pathValue });
```

Comment:
Some comments on what this is all about would be nice, right?  (It's probably my fault, so hopefully I remember correctly).

The `OutputType` on cmdlets like `Get-ChildItem` may depend on the path. The `CmdletInfo` returned based on just the command name will specify returning all possibilities, e.g. certificates, environment, registry, etc.

If you specified `-Path`, the list of `OutputType` can be refined, but we have to make a copy of the `CmdletInfo` object this way to get that refinement.